subsys/net/l2/wifi/wifi_shell.c

```diff
@@ -1141,14 +1141,33 @@ static int cmd_wifi_ap_disable(const struct shell *sh, size_t argc,
 static int cmd_wifi_reg_domain(const struct shell *sh, size_t argc,
 			       char *argv[])
 {
+#if (CONFIG_HEAP_MEM_POOL_SIZE > 0)
 	struct net_if *iface = net_if_get_first_wifi();
-	struct wifi_reg_domain regd = {0};
-	int ret;
+	struct wifi_reg_domain *regd = NULL;
+	int ret, i = 0, len = 0;
+
+	len = sizeof(struct wifi_reg_domain);
 
 	if (argc == 1) {
-		regd.oper = WIFI_MGMT_GET;
+		len += sizeof(struct wifi_reg_chnl_info) * MAX_REG_CHNL_NUM;
+		regd = k_malloc(len);
```

Comment:
Where is this memory freed?

Could we avoid using dynamic memory as it is very easy to leak memory that way?
Could the data be "compressed" so transfer only data that is different from default etc?
Could we use a static buffer and pass pointer to that instead?

I am asking these questions because we just fixed a memory hog issue in  #62584 few months ago and I would like to avoid similar issue with this one.

Reply:
(1) Freeing memory is missed, I will add up. 

(2) Could the data be "compressed" so transfer only data that is different from default etc?
There is no default information, it's country specific information so it will vary from one to other . Like some channels are supported in world regulatory but not supported in US. Each country/regulatory domain has  it's own set of values and it keeps changing.  
Please update more about "transfer only data that is different from default". 

(3) Could we use a static buffer and pass pointer to that instead?
Static buffer means using earlier one "struct wifi_reg_chn_info chnl_info[MAX_REG_CHNL_NUM];" or something else.
I was passing this pointer earlier too. Is that one you are mentioning here or something else.    

I can do compresssion of the structure as you suggested below, so that overall static memory will be reduced, but still not sure if that is sufficient. It will be reduced to say 250 bytes. But if we are going to support 60MHZ more channels ord other info also will be added then it will again be an issue.

Reply:
Isn't the reg_domain more or less static meaning it does not change much?
So could we have a static variable and we just pass pointer value between the shell and the driver. The shell / driver would need to copy the data somewhere in this case too. This way we would avoid any memory allocation issues and could save some memory.

Reply:
> Isn't the reg_domain more or less static meaning it does not change much? So could we have a static variable and we just pass pointer value between the shell and the driver. The shell / driver would need to copy the data somewhere in this case too. This way we would avoid any memory allocation issues and could save some memory.

We can also change the regulatory domain dynamically from the shell through "regd->oper = WIFI_MGMT_SET" at any time. Regulatory data comes from firmware. Driver does not not maintain any regulatory information.  Once the firmware generates event, through the driver data is copied to shell/application memory.